src/Http/Routing/src/Matching/ILEmitTrieFactory.cs

```diff
@@ -102,25 +102,17 @@ internal static bool ShouldVectorize((string text, int destination)[] entries)
         // ref byte p = ...
         il.Emit(OpCodes.Stloc_0, locals.P);
 
+        const int binarySearchThreshold = 4; // The number of items above which it makes sense to binary search
```

Comment:
Might be worth mentioning how this was determined.

Reply:
I arrived at this number after doing some analysis, described below. Consider that as we are emitting the code essentially "hardcoding" the tree into IL, we are able to use this heuristic to determine what is a good binary tree size.

In the table below, we analyze how the linear vs binary search tree performance varies depending on number of items, in terms of number of equality checks (`==`).
 - The BST approach implies first doing a comparison operation (`>`) to split the number of items into two buckets followed immediately by a linear search.
 - This comparison operation for BST should be kept in mind when reading the table.
 - We analyze the AverageCase scenario: number of `==` ops if all items access once, divided by number of items.
 - We analyze the WorstCase scenario: the largest number of `==` ops possible

| Items | Strategy    | AverageCase | WorstCase: |
| ----- | ----------- | ----------- | ---------- |
|     2 | Linear      |         1.5 |          2 |
|     2 | BST         |           1 |          1 |
|     3 | Linear      |           2 |          3 |
|     3 | BST         |         1.6 |          2 |
|     4 | Linear      |           5 |          4 |
|     4 | BST         |           2 |          2 |
|     5 | Linear      |         7.5 |          5 |
|     5 | BST         |         2.8 |          3 |

Based on the table above, and keeping in mind that the BST approach **always implies a single extra comparison/branch step**, we conclude:
1. For 2 and 3 items Linear is better in both Average and Worst Cases (when considering the extra comparison/branch)
2. For 4 items, I decided that the cost of the extra comparison/branch is worth the reduction of AverageCase from 5 -> 2 and WrostCase from 4 -> 2.
3. For 5+ items, the benefit starts to become much clearer.

Reply:
Not sure if i'm reading this the wrong way, but I can't understand how the average case can be worst than the worst case for linear comparisons with values 4 and 5.

Reply:
On review, it's clear to me that I've divided by 2 instead of `count`. There were a few other mistakes too which I have updated alongside my working for transparency.

I've updated the table and added my working to regain some credibility I guess 🤷